### ISSUE TITLE
SMB is broken - gone from the list of available external storages

### ISSUE THREAD

**Issue by mmattel** [owncloud/core#15546]
@icewind1991 @MorrisJobke 

Ubuntu 14.04, MySQL, nginx

Something bad happend.

After merging  #15535, I rebased my dev environment today morning and saw that SMB has completely gone from the available external storage list. After resetting my dev environment and reloding it from git with ``ocdev setup base``, enabling files_external, SMB was still missing in the list of available storages.

This is how the list in the admin page looks like now:
![image](https://cloud.githubusercontent.com/assets/3321281/7101234/02efb326-e04f-11e4-8ffa-dbd15acf3ec7.png)

The patches applied in the SMB code as reason for the merge refrenced do not seem to be the cause, as this has worked before, but maybe registering of SMB itself causes the problem.

Following logs:

nginx does not have any corresponding error log entries for this time.

owncloud log:

     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#282
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#281
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#280
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#279
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#278
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#277
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#276
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#275
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#274
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#273
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#272
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#271
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#270
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#269
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#268
     Trying to get property of non-object at /xxx/lib/private/ocsclient.php#267

code lines affected at ocsclient.php:
starting with line 267

                $app = [];
                $app['id'] = (int)$tmp->id;
                $app['name'] = (string)$tmp->name;
                $app['version'] = (string)$tmp->version;
                $app['type'] = (string)$tmp->typeid;
                $app['label'] = (string)$tmp->label;
                $app['typename'] = (string)$tmp->typename;
                $app['personid'] = (string)$tmp->personid;
                $app['profilepage'] = (string)$tmp->profilepage;
                $app['detailpage'] = (string)$tmp->detailpage;
                $app['preview1'] = (string)$tmp->smallpreviewpic1;
                $app['preview2'] = (string)$tmp->smallpreviewpic2;
                $app['preview3'] = (string)$tmp->smallpreviewpic3;
                $app['changed'] = strtotime($tmp->changed);
                $app['description'] = (string)$tmp->description;
                $app['detailpage'] = (string)$tmp->detailpage;
                $app['score'] = (int)$tmp->score;

**Comment by karlitschek** [comment]
@icewind1991 Can you have a look? Is the smclient now a hard dependency?

**Comment by DeepDiver1975** [comment]
I just had a look on current master and SMB is listed

**Comment by mmattel** [comment]
hmmm... I just tried it again, and it is still not listed.
disabling and re-enabeling files_external, smb is still not present - what can I do to provide further information?

**Comment by DeepDiver1975** [comment]
```php
	/**
	 * check if smbclient is installed
	 */
	public static function checkDependencies() {
		$smbClientExists = (bool)\OC_Helper::findBinaryPath('smbclient');
		return $smbClientExists ? true : array('smbclient');
	}

**Comment by DeepDiver1975** [comment]
the smb external storage implementation is still checking for smbclient being installed - but in case the native implementation of libsmbclient-php

**Comment by DeepDiver1975** [comment]
@mmattel I assume smbclient is not installed on your system?

**Comment by mmattel** [comment]
I just changed the code in checkDependencies to return always true for testing.
Having this patched, it works now.

I do not know what else / additionally changed up to #15535 was merged, but it worked before...

**Comment by mmattel** [comment]
No smbclient (also libsmbclient) was and is installed on my system.

    sudo apt-get install smbclient
    Reading package lists... Done
    Building dependency tree
    Reading state information... Done
    smbclient is already the newest version.

**Comment by mmattel** [comment]
which smbclient
    /usr/bin/smbclient

It seems, hat \OC_Helper::findBinaryPath('smbclient') does not return a value

**Comment by karlitschek** [comment]
@DeepDiver1975 @icewind1991 So cifs support has now a hard requirement on smclient? I thought the "old" PHP way is still supported?

**Comment by DeepDiver1975** [comment]
> So cifs support has now a hard requirement on smclient?

No/Yes - this is the old way - which we still support.

The dependency check is missing the situation where the native php module is enabled - we shall add this

**Comment by mmattel** [comment]
Just tested by replacing in the query ``smclient`` with eg ``sendmail`` which is definitely present - same error.
``\OC_Helper::findBinaryPath`` does not return a result here?

**Comment by MorrisJobke** [comment]
cc @nickvergessen Didn't you changed something in there?

**Comment by karlitschek** [comment]
@DeepDiver1975 Ahh got it. Sometimes it makes sense to look into the code before asking stupid questions ;-)

**Comment by nickvergessen** [comment]
I just changed it so it does not run on windows.
However @oparoz modfied the paths we look at https://github.com/owncloud/core/pull/12314: https://github.com/owncloud/core/commit/0f01de9f7e3a17afa769680f25f73ddef15b1ec0

The other changes should be save:
https://github.com/owncloud/core/commit/c29138311614cba8bc4699a0afa485aefe621b6b https://github.com/owncloud/core/commit/7c41c0c13f1a845fdcedd62218de490e8c01691b

**Comment by oparoz** [comment]
`findBinaryPath` is using Symfony's `ExecutableFinder`, but even when that fails, `find` should return a result.

@mmattel What do you get when running the `find` command from the shell?
https://github.com/oparoz/core/blob/7c41c0c13f1a845fdcedd62218de490e8c01691b/lib/private/helper.php#L929

**Comment by nickvergessen** [comment]
@oparoz you changed the paths `find` looks at, so I assume that is the problem :wink:

**Comment by mmattel** [comment]
find /usr/local/bin /usr/bin/ -name smbclient 2> /dev/null
    /usr/bin/smbclient

**Comment by oparoz** [comment]
@nickvergessen - . I didn't change the paths, I replaced `which` with `ExecutableFinder` which should be more reliable, even though it still doesn't work properly for people with open_basedir restrictions.

`find` is only used as a backup solution and it's looking in /usr/bin, so something should be returned. 

@mmatel - Thanks
Try to comment out https://github.com/oparoz/core/blob/7c41c0c13f1a845fdcedd62218de490e8c01691b/lib/private/helper.php#L919-L921 to see if you get a result

**Comment by mmattel** [comment]
I commented out

		if ($memcache->hasKey($program)) {
			return $memcache->get($program);
		}

But the problem is still there

**Comment by oparoz** [comment]
@mmattel - Thanks
If it's easy for you to do, it would be great if you could log the output of
https://github.com/oparoz/core/blob/7c41c0c13f1a845fdcedd62218de490e8c01691b/lib/private/helper.php#L926 to see if you get a bogus result from `ExecutableFinder`

and also, commenting out lines 930 and 932 to log the output of line 931.

**Comment by mmattel** [comment]
@oparoz 
Info the code updated from git on my machine (updated today) compared to the linked code does not match exactly.

Here is what I did:

                $result = null;
                if (!\OC_Util::runningOnWindows() && self::is_function_enabled('exec')) {
                        $exeSniffer = new ExecutableFinder();
                        // Returns null if nothing is found
                        $result = $exeSniffer->find($program);
                        \OCP\Util::writeLog('program', $program, \OCP\Util::DEBUG);
                        \OCP\Util::writeLog('test 1', $result, \OCP\Util::DEBUG);
                        if (empty($result)) {
                                $paths = str_replace(':','/ ',getenv('PATH'));
                                $command = 'find ' . $paths . ' -name ' . escapeshellarg($program) . ' 2> /dev/null';
                                exec($command, $output, $returnCode);
                        \OCP\Util::writeLog('test 2', $command, \OCP\Util::DEBUG);

And here is the result:

    Debug	test 2	find -name 'smbclient' 2> /dev/null	2015-04-13T11:52:51+00:00
    Debug	test 1		                                2015-04-13T11:52:51+00:00
    Debug	program	smbclient	                        2015-04-13T11:52:51+00:00

It seems, that the path creation routine does not do the job properly

**Comment by mmattel** [comment]
Added a log write to see if the getenv('PATH') returns a value.
Interestingly there is no output??
But if I do following from the php command line, I get:

    php > echo getenv('PATH');
    /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/gamesphp >

**Comment by oparoz** [comment]
OK, so it seems there is a problem on your setup with `getenv`. `ExecutableFinder` is using the same command to get the paths it can search in.

Could you add a dummy page to oC, serving phpinfo(), to see if you get any paths at all?

**Comment by nickvergessen** [comment]
@oparoz https://github.com/owncloud/core/commit/0f01de9f7e3a17afa769680f25f73ddef15b1ec0 yes you changed it ;)

**Comment by oparoz** [comment]
@nickvergessen - Ah, I see what you mean now. I was asked to do it that way:
https://github.com/owncloud/core/pull/12314#issuecomment-76203388

but that's not really the problem here. `getenv('PATH')` should return a list of paths.

**Comment by nickvergessen** [comment]
yes, but its a different list

**Comment by oparoz** [comment]
Without a working `getenv('PATH')`other areas of oC will break
https://github.com/owncloud/core/blob/3d2116dfe750bceecdde464985046b3116e3e470/lib/private/helper.php#L565

And if `getenv()` is broken, there are even more problems.

**Comment by oparoz** [comment]
@mmattel - Do you get something if you use `$_SERVER["PATH"]` instead?

**Comment by mmattel** [comment]
I quickly created a browser access to phpinfo() and the result was:
![image](https://cloud.githubusercontent.com/assets/3321281/7116484/2288ce8a-e1f2-11e4-8046-6b8d700400ec.png)

``$_SERVER["PATH"]`` drops a php error

Then I did some research and found that when you use php5-fpm, the environment variables are NOT automatically beeing populated by php. I took the remark off in ``sudo vi /etc/php5/fpm/pool.d/www.conf``

    ; Pass environment variables like LD_LIBRARY_PATH. All $VARIABLEs are taken from
    ; the current environment.
    ; Default Value: clean env
    ;env[HOSTNAME] = $HOSTNAME
    env[PATH] = /usr/local/bin:/usr/bin:/bin
    ;env[TMP] = /tmp
    ;env[TMPDIR] = /tmp
    ;env[TEMP] = /tmp

![image](https://cloud.githubusercontent.com/assets/3321281/7116835/fe947940-e1f4-11e4-8490-9f8dbaf79208.png)

And the function returned a proper value letting SMB appear again

**Comment by oparoz** [comment]
OK. So @nickvergessen , @DeepDiver1975, maybe a working getenv('PATH') should be added as a dependency somehow?

**Comment by nickvergessen** [comment]
@LukasReschke as dependency check master

**Comment by mmattel** [comment]
I have searched the code for getenv ``server variables`` queries and found:

lib/base.php (L1103+)

	if ($temp = getenv('TMP')) return $temp;
	if ($temp = getenv('TEMP')) return $temp;
	if ($temp = getenv('TMPDIR')) return $temp;

lib/private/helper.php (L565 and L934)

	$path = getenv("PATH");
	$paths = str_replace(':','/ ',getenv('PATH'));

We need to document this dependency.

And in addition I propose the following:

We do eg checks and write out the status of these checks at the admin page at the top at ``Security and Setup Warnings``. Can this be warned here too?? This is so easy to overseeable that a warning here is imho necessary.

**Comment by nickvergessen** [comment]
the `lib/base.php` checks are save, since they fallback to other values if they dont exist.

**Comment by oparoz** [comment]
I'm happy to add yet another fallback to `findBinaryPath` to simplify things for admins, hoping that `/usr/bin` and `/usr/local/bin` cover 80% of uses.

**Comment by mmattel** [comment]
That would be great.
If getenv('PATH') returns no result, take a default like above. This definitely would minimize problems.

**Comment by oparoz** [comment]
@mmattel - Could you test #15597?

**Comment by DeepDiver1975** [comment]
fixed with #15597
